ci-operator/step-registry/cucushift/upgrade/toversion/cucushift-upgrade-toversion-commands.sh

```diff
@@ -0,0 +1,487 @@
+    fi
+
+    echo "Require admin ack:\n ${out}"
+    local wait_time_loop_var=0 ack_data
+
+    ack_data="$(echo "${out}" | jq -r "keys[]")"
+    for ack in ${ack_data};
+    do
+        # e.g.: ack-4.12-kube-1.26-api-removals-in-4.13
+        if [[ "${ack}" == *"ack-4.${SOURCE_MINOR_VERSION}"* ]]
+        then
+            echo "Admin ack patch data is: ${ack}"
+            oc -n openshift-config patch configmap admin-acks --patch '{"data":{"'"${ack}"'": "true"}}' --type=merge
+        fi
+    done
+    echo "Admin-acks patch gets started"
+
+    echo -e "sleep 5 min wait admin-acks patch to be valid...\n"
+    while (( wait_time_loop_var < 5 )); do
+        sleep 1m
+        echo -e "wait_time_passed=${wait_time_loop_var} min.\n"
+        if ! oc adm upgrade | grep "AdminAckRequired"; then
+            echo -e "Admin-acks patch PASSED\n"
+            return 0
+        else
+            echo -e "Admin-acks patch still in processing, waiting...\n"
+        fi
+        (( wait_time_loop_var += 1 ))
+    done
+    if (( wait_time_loop_var >= 5 )); then
+        echo >&2 "Timed out waiting for admin-acks completing, exiting" && return 1
+    fi
+}
+
+# check if any of cases is enabled via ENABLE_OTA_TEST
+function check_ota_case_enabled() {
+    local case_id
+    local cases_array=("$@")
+    for case_id in "${cases_array[@]}"; do
+        # shellcheck disable=SC2076
+        if [[ " ${ENABLE_OTA_TEST} " =~ " ${case_id} " ]]; then
+            echo "${case_id} is enabled via ENABLE_OTA_TEST on this job."
+            return 0
+        fi
+    done
+    return 1
+}
+
+function set_channel(){
+    x_ver=$( echo "${DUMMY_TARGET_VERSION}" | cut -f1 -d. )
+    y_ver=$( echo "${DUMMY_TARGET_VERSION}" | cut -f2 -d. )
+    ver="${x_ver}.${y_ver}"
+    if ! oc adm upgrade channel candidate-${ver}; then
+        echo "Fail to change channel to candidate-${ver}!"
+        return 1
+    fi
+    local retry=3
+    while (( retry > 0 ));do
+        recommends=$(oc get clusterversion version -o json|jq -r '.status.availableUpdates')
+        if [[ "${recommends}" == "null" ]]; then
+            (( retry -= 1 ))
+            sleep 60
+            echo "No recommended update available! Retry..."
+        else
+            echo "Recommencded update: ${recommends}"
+            break
+        fi
+    done
+    if (( retry == 0 )); then
+        echo "Timeout to get recommended update!" 
+        return 1
+    fi
+}
+
+# judge if upgrade with --to or --to-latest cmd
+# set correct target_version for the upgrade
+function set_target_and_upgrade_cmd(){
+    mapfile -t available_versions < <(echo "$(oc get clusterversion version -o json|jq -r '.status.availableUpdates[]?.version')")
+    if [[ -z "${available_versions[*]}" ]]; then
+        echo "No recommended update available!"
+        exit 1
+    fi
+    # for "oc adm upgrade --to x.y.z" and "oc adm upgrade --to $latest_version_in_available_update"
+    upgrade_cmd="oc adm upgrade --to ${TARGET_VERSION}"
```

Comment:
I did not not see where defined the `TARGET_VERSION` before referenced it in this command line.

Reply:
yeah, updated